### PR TITLE
Fix `[semver]` type to pass semver.org tests

### LINKED ISSUE #21383: `[semver]` allows invalid version number `1.0.0-beta..1`, not allowed by semver.org suggested regex
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

Here are an example of a invalid semantic versioning version number that is not valid by semver.org suggested regex:

* <https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string>

```powershell
## Should work
[semver]'1'
[semver]'1.0.0'
[semver]'1.0.0-beta'
[semver]'1.0.0-beta1'
[semver]'1.0.0-beta.1'

## Should error but doesn't
[semver]'1.0.0-beta..1'
[semver]'1.0.0-beta...1'

## Should error and does
[semver]'1.0.0-beta!1'
```

This tells me `[semver]` isn't fully compliant with [semantic versioning v2.0.0](https://semver.org/spec/v2.0.0.html).

### Expected behavior

```console
Follow semver specification. Throw an error on non-valid version numbers.
```

### Actual behavior

```console
PS > [semver]'1.0.0-beta...1'

Major  Minor  Patch  PreReleaseLabel BuildLabel
-----  -----  -----  --------------- ----------
1      0      0      beta...1

PS >
```

### Error details

_No response_

### Environment data

```powershell
PowerShell v7.4.1 x64
Windows 11 23H2 x64
```

### Visuals

_No response_

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
# PR Summary

- replaced existing regex with slightly modified one from semver.org matching C# syntax
- added separate regex for buildLabel validation as it differs from the preLabel validation
- Fix some misspellings of `semver` in the comments
- Added all the valid and invalid tests from semver.org except the one where the version numbers exceed an int

## PR Context

Fix https://github.com/PowerShell/PowerShell/issues/21383

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: src/System.Management.Automation/engine/PSVersionInfo.cs [hunk 1]
```diff
@@ -336,7 +336,7 @@ public sealed class SemanticVersion : IComparable, IComparable<SemanticVersion>,
     {
         private const string VersionSansRegEx = @"^(?<major>\d+)(\.(?<minor>\d+))?(\.(?<patch>\d+))?$";
         private const string LabelRegEx = @"^((?<preLabel>[0-9A-Za-z][0-9A-Za-z\-\.]*))?(\+(?<buildLabel>[0-9A-Za-z][0-9A-Za-z\-\.]*))?$";
-        private const string LabelUnitRegEx = @"^[0-9A-Za-z][0-9A-Za-z\-\.]*$";
+        private const string LabelUnitRegEx = @"^[0-9A-Za-z-][0-9A-Za-z\-\.]*$";
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/PSVersionInfo.cs:339]: Please add the same update to `LabelRegEx` too. 
And new test for constructor using `LabelRegEx`.

**Review by iSazonov — COMMENTED** [review]
The PR doesn't fix  #21383 - we still pass [semver]'1.0.0-beta..1'

I do not understand the discussion of this class and attempts to fix it, because the only purpose of adding it was to ensure the pwsh release process and nothing more, and the class itself was always considered `psevdo-public, for internal use only`.
It might be worth adding it to the documentation if it hasn't been done yet.

File: src/System.Management.Automation/engine/PSVersionInfo.cs (hunk 1 above)
Comment (SteveL-MSFT) [reply src/System.Management.Automation/engine/PSVersionInfo.cs]: Looks like this code was incorrectly using `LabelUnitRegEx` to validate both the prerelease and build labels.  Updated the regex's to match the ones from semver.org (complying with C# syntax).  Fixed the preLabel regex as it's actually using the buildLabel regex previously and added new buildLabel regex.  Fixed the tests so they should be validating correctly now.

**Review by iSazonov — APPROVED** [review]

**Review by SeeminglyScience — APPROVED** [review]
LGTM!

**Pull request merged**